src/lib/log.c

```diff
@@ -19,19 +19,94 @@
 
 #define _GNU_SOURCE
 
+#include <ctype.h>
 #include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
 #include <time.h>
 
 #include "log.h"
 #define DEBUG_HEADER_SIZE 50
 
 static int cur_log_level = LOG_INFO;
+static int json_format = 0;
 
 void set_log_level(int log_level)
 {
 	cur_log_level = log_level;
 }
 
+void set_json_format(void)
+{
+	json_format = 1;
+}
+
+void json_start(const char *op)
+{
+	if (json_format) {
+		fprintf(stderr, "[\n");
+		fflush(stderr);
+		fprintf(stderr, "{ \"%s\" : \"start\" },\n", op);
+		fflush(stderr);
+	}
+}
+
+void json_end(const char *op)
+{
+	if (json_format) {
+		fprintf(stderr, "{ \"%s\" : \"end\" }\n", op);
+		fflush(stderr);
+		fprintf(stderr, "]\n");
+		fflush(stderr);
+	}
+}
+
+void format_to_json(const char *type, const char *msg, va_list args_list)
+{
+	const int MAX_TYPE_LENGTH = 8;
+	char *full_msg;
+	char ltype[MAX_TYPE_LENGTH];
+
+	/* make sure the type is all lower case */
```

Comment:
It looks like there's no easy way to perform a toLower in a string in c :(
This is one thing that could move to one specific function, like str_lower(). You can create the string using malloc(strlen(string)) and return that, so we don't need to assume there's a max length.

Reply:
ok, done